File: c#-solution/0205. Isomorphic Strings.cs
// 205. Isomorphic Strings
// https://leetcode.com/problems/isomorphic-strings/description/
// Tag: C#, HashTable, String, Easy

public class Solution {
    public bool IsIsomorphic(string s, string t) {
        if(s.Length != t.Length)return false;
        var map1 = new Dictionary<char, char>();
        var map2 = new Dictionary<char, char>();

        for(int i=0; i<s.Length; i++){
            if(!map1.ContainsKey(s[i])){
                map1.Add(s[i], t[i]);
            } else {
                if(map1[s[i]] != t[i]) return false;
            }
            if(!map2.ContainsKey(t[i])){
                map2.Add(t[i], s[i]);
            } else {
                if(map2[t[i]] != s[i]) return false;
            }
        }
        return true;
    }
}
// TC: O(n)
// SC: O(n)

File: problem/0205. Isomorphic Strings.js
// 205. Isomorphic Strings
// https://leetcode.com/problems/isomorphic-strings/
// Tag: String, Hash Table, Easy

// Given two strings s and t, determine if they are isomorphic.
// Two strings s and t are isomorphic if the characters in s can be replaced to get t.

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isIsomorphic = function(s, t) {
    const maps = new Map();
    const mapt = new Map();
    for(let i=0; i<s.length; i++){
        if(!maps.has(s[i])){
            maps.set(s[i], t[i]);
        } else {
            if(maps.get(s[i]) !== t[i]) return false;
        }
        if(!mapt.has(t[i])){
            mapt.set(t[i], s[i]);
        } else {
            if(mapt.get(t[i]) != s[i]) return false;
        }
    }
    return true;
};

// -- for testing --
console.log(isIsomorphic("egg", "add")); // true
console.log(isIsomorphic("foo", "bar")); // false